File: crypto-analyzer/core/orchestrator.py
import asyncio

from apps.analytics.crud import get_actual_components, get_needed_fields_from_endpoint, add_trigger_history
from apps.analytics.gateway_router import send_triggered_formulas
from apps.binance.external_router import BinanceAPI
from core.analysis.graph import dependency_graph
from core.logger import custom_logger


class BinanceAPIOrchestrator:
    """
    запускает периодические задачи
    на первый взгляд выглядит не плохо, вот только большое количество тасок он сейчас не вывезет
    """
    def __init__(self, binance_api: BinanceAPI):
        self.binance_api = binance_api
        self.is_binance_online = True
        self.tasks: dict[str, asyncio.Task] = {} # лучше это не менять
        self.task_cooldowns: dict[str, int] = {}

    async def start(self):
        """Запуск фоновых задач. первая задача должна быть проверка апи"""
        response = await self.binance_api.get_apiv3_accessibility()
        await self.check_binance_response(response)

        if self.is_binance_online:
            await self.launch_needed_api()

    async def launch_needed_api(self) -> None:
        """выбирает апи, которые нужны для получения актуальных данных в граф, и убирает ненужные"""
        print("launching needed API tasks...")
        data = await get_actual_components()
        current_apis = set(data.keys())
        running_apis = set(self.tasks.keys()) - {"weights"}

        outdated_apis = running_apis - current_apis
        for api in outdated_apis:
            print(f"stopping outdated API: {api}")

            task = self.tasks.pop(api, None)
            self.task_cooldowns.pop(api, None)

            if task and not task.done():
                task.cancel()
            else:
                custom_logger.log_with_path(
                    level=3,
                    msg=f"{type(task), task, self.tasks}",
                    filename="StrangeError.log"
                )

        for api, (cooldown, formulas_amount) in data.items():
            if formulas_amount <= 0:
                continue

            existing_task = self.tasks.get(api)
            if existing_task and not existing_task.done():
                continue

            print(f"Starting API task: {api} (cooldown: {cooldown}seconds)")
            self.launch_api(api, cooldown)

    async def adjust_api_task_cooldown(self, api: str, new_cooldown: int) -> None:
        """Updates the frequency of proxification for a particular API, if necessary"""
        if api not in self.tasks:
            return

        old_cooldown = self.task_cooldowns.get(api)
        if old_cooldown == new_cooldown:
            return

        task = self.tasks.pop(api)
        self.task_cooldowns.pop(api)
        if not task.done():
            task.cancel()
        self.launch_api(api, new_cooldown)

    def launch_api(self, api: str, cooldown: int) -> None:
        if api == "/v3/ticker/price":
            self.tasks[api] = asyncio.create_task(self.update_ticker_current_price(cooldown=cooldown))
            self.task_cooldowns[api] = cooldown
        elif api == "/v3/ticker/24hr":
            self.tasks[api] = asyncio.create_task(self.update_price_change_24h(cooldown=cooldown))
            self.task_cooldowns[api] = cooldown
        elif api == "/v3/ping":
            self.tasks[api] = asyncio.create_task(self.check_api_status_loop(cooldown=cooldown))
            self.task_cooldowns[api] = cooldown

    async def check_binance_response(self, response):
        """checks Binacne availability"""
        if not isinstance(response, dict) and not isinstance(response, list) or response is True: # бинанс лег
            if self.is_binance_online is True:  # первое срабатывание - документация
                custom_logger.log_with_path(
                    level=3,
                    msg=f"Binance DOWN",
                    filename="BinanceAvailablity.log"
                )
                self.is_binance_online = False
            if len(self.tasks) > 1:
                for task in self.tasks:
                    task.cancel()
                self.launch_api("/v3/ping", 60)
        elif self.is_binance_online is False:  # бинанс ожил
            custom_logger.log_with_path(
                level=3,
                msg=f"Binance UP",
                filename="BinanceAvailablity.log"
            )
            self.is_binance_online = True
            for task in self.tasks:
                if isinstance(task, asyncio.Task):
                    task.cancel()
            await self.launch_needed_api()

    async def check_api_status_loop(self, cooldown: int):
        """Проверка доступности API."""
        while True:
            await asyncio.sleep(cooldown)
            response = await self.binance_api.get_apiv3_accessibility()
            await self.check_binance_response(response)
            if self.is_binance_online is True:
                break

    async def update_ticker_current_price(self, cooldown: int):
        while True:
            print('update_ticker_current_price')
            await asyncio.sleep(cooldown)
            response = await self.binance_api.get_ticker_current_price()
            await self.check_binance_response(response)
            currencies = await get_needed_fields_from_endpoint(endpoint="/v3/ticker/price")

            data_for_graph = extract_data_from_ticker_current_price(response, currencies)
            triggered_formulas = dependency_graph.update_variables_topological_Kahn(data_for_graph)
            if len(triggered_formulas) > 0:
                result, variable_values = dependency_graph.get_formulas_variables(triggered_formulas)
                await add_trigger_history(result, variable_values)

                await send_triggered_formulas(formulas=triggered_formulas)

    async def update_price_change_24h(self, cooldown: int):
        while True:
            print('update_price_change_24h')
            await asyncio.sleep(cooldown)
            response = await self.binance_api.get_price_change_24h()
            await self.check_binance_response(response)
            fields = await get_needed_fields_from_endpoint(endpoint="/v3/ticker/24hr")

            data_for_graph = extract_data_from_price_change_24h(response, fields)
            triggered_formulas = dependency_graph.update_variables_topological_Kahn(data_for_graph)
            if len(triggered_formulas) > 0:
                result, variable_values = dependency_graph.get_formulas_variables(triggered_formulas)
                await add_trigger_history(result, variable_values)

                await send_triggered_formulas(formulas=triggered_formulas)


def extract_data_from_price_change_24h(dataset: list, fields: dict[str, list]) -> dict:
    dataset_dict = {data["symbol"]: data for data in dataset}
    result = {}

    for symbol, field_list in fields.items():
        if symbol in dataset_dict:
            data = dataset_dict[symbol]
            for field in field_list:
                if field in data:
                    result[f"{symbol}_{field}"] = float(data[field])

    return result

def extract_data_from_ticker_current_price(dataset: list, currencies: dict[str, list[str]]) -> dict:
    """тут только symbol и price"""
    dataset_dict = {data["symbol"]: data for data in dataset}
    result = {}

    for symbol in currencies.keys():
        data = dataset_dict.get(symbol)
        if data:
            result[f"{symbol}_price"] = float(data["price"])

    return result

File: crypto-gateway/internal/analytics/orchestrator.go
package analytics

import (
	"context"
	"crypto-gateway/internal/appmetrics"
	"crypto-gateway/internal/web/repositories"
	"encoding/json"
	"fmt"
	"log"
	"strconv"
	"sync"
	"time"
)

// отвечает за запуск периодических задач
type BinanceAPIOrchestrator struct {
	DependencyGraph *DependencyGraph
	binanceAPI      *BinanceAPI
	isBinanceOnline bool

	mu            sync.Mutex
	tasks         map[string]context.CancelFunc
	taskCooldowns map[string]int
}

func NewBinanceAPIOrchestrator(api *BinanceAPI) *BinanceAPIOrchestrator {
	return &BinanceAPIOrchestrator{
		DependencyGraph: NewDependencyGraph(),
		binanceAPI:      api,
		isBinanceOnline: getBinanceStatusOnStartUp(),
		tasks:           make(map[string]context.CancelFunc),
		taskCooldowns:   make(map[string]int),
	}
}

// определеяет начальное значение переменной isBinanceOnline и устанавливает начальные логи бинанса
func getBinanceStatusOnStartUp() bool {
	response, err := StBinanceApi.Get(context.Background(), "/v3/ping", endpoints["/v3/ping"], nil)
	var data map[string]interface{}
	err2 := json.Unmarshal(response, &data)

	if err == nil && err2 == nil && len(data) == 0 {
		fmt.Println("Binance UP")
		appmetrics.AvailabilityMetricEvent(2, 1, "Binance UP")
		return true
	} else {
		fmt.Println("Binance DOWN")
		appmetrics.AvailabilityMetricEvent(2, 0, "Binance DOWN")
		return false
	}

	/*
		// относительно менее надежная проверка
		if err != nil { // Binance DOWN
			fmt.Println("Binance DOWN")
			AvailabilityMetricEvent(2, 0, "Binance DOWN")
			return false
		} else { // Binance UP
			fmt.Println("Binance UP")
			AvailabilityMetricEvent(2, 1, "Binance UP")
			return true
		}
	*/
}

// Запуск фоновых задач. первая задача должна быть проверка доступности апи бинанса
func (o *BinanceAPIOrchestrator) Start(ctx context.Context) error {
	if o.isBinanceOnline {
		o.checkBinanceResponse([]int{})
	} else {
		o.checkBinanceResponse(nil)
	}

	o.LoadActiveFormulas(ctx)
	if o.isBinanceOnline {
		o.LaunchNeededAPI(ctx)
	}

	return nil
}

// выбирает формулы из trigger_formula у которых is_active=true и добавляет их в граф
func (o *BinanceAPIOrchestrator) LoadActiveFormulas(ctx context.Context) {
	fmt.Println("Loading active formulas in Graph...")

	formulas, err := GetActiveFormulas(ctx)
	if err != nil {
		log.Printf("ошибка загрузки активных формул: %v\n", err)
		return
	}

	for _, f := range formulas {
		if err := o.DependencyGraph.AddFormula(f.Formula, f.ID); err != nil {
			appmetrics.ApplicationCriticalErrorsLogging(fmt.Sprintf("Failed to add formula %v with id %v to graph in LoadActiveFormulas", f.Formula, f.ID), err)
		}
	}
	fmt.Println("Graph построен для активных формул.")
}

// выбирает апи, которые нужны для получения актуальных данных в граф, и убирает ненужные
func (o *BinanceAPIOrchestrator) LaunchNeededAPI(ctx context.Context) {
	fmt.Println("Launching needed API tasks...")

	data, err := GetActualComponents(ctx)
	fmt.Println(data)
	if err != nil {
		appmetrics.ApplicationCriticalErrorsLogging("GetActualComponents returned error in LaunchNeededAPI function", err)
	}

	o.mu.Lock()
	defer o.mu.Unlock()

	currentAPIs := make(map[string]struct{})
	for k := range data {
		currentAPIs[k] = struct{}{}
	}

	runningAPIs := make(map[string]struct{})
	for k := range o.tasks {
		if k != "weights" {
			runningAPIs[k] = struct{}{}
		}
	}

	// остановка устаревших задач
	for api := range runningAPIs {
		if _, ok := currentAPIs[api]; !ok {
			fmt.Printf("Stopping outdated API: %s\n", api)
			cancelFunc := o.tasks[api]
			cancelFunc()
			delete(o.tasks, api)
			delete(o.taskCooldowns, api)
		}
	}

	// запуск новых, в следствии обновления периодичности или создания формул которым нужны данные других апи
	for api, info := range data {
		if info.Count <= 0 {
			continue
		}

		if _, exists := o.tasks[api]; exists {
			continue
		}

		fmt.Printf("Starting API task: %s (cooldown: %d seconds)\n", api, info.Cooldown)
		o.launchAPI(ctx, api, info.Cooldown)
	}
}

// Updates the frequency of proxification for a particular API, if necessary
func (o *BinanceAPIOrchestrator) AdjustAPITaskCooldown(ctx context.Context, api string, newCooldown int) {
	o.mu.Lock()
	defer o.mu.Unlock()

	oldCooldown, exists := o.taskCooldowns[api]
	if !exists || oldCooldown == newCooldown {
		return
	}

	cancelFunc, exists := o.tasks[api]
	if exists {
		cancelFunc()
		delete(o.tasks, api)
		delete(o.taskCooldowns, api)
	}

	o.launchAPI(ctx, api, newCooldown)
}

func (o *BinanceAPIOrchestrator) launchAPI(ctx context.Context, api string, cooldown int) {
	cctx, cancel := context.WithCancel(ctx)
	o.tasks[api] = cancel
	o.taskCooldowns[api] = cooldown

	switch api {
	case "/v3/ticker/price":
		go o.updateTickerCurrentPrice(cctx, cooldown)
	case "/v3/ticker/24hr":
		go o.updatePriceChange24h(cctx, cooldown)
	case "/v3/ping":
		go o.checkAPIStatusLoop(cctx, cooldown)
	default:
		fmt.Printf("Unknown API task: %s\n", api)
	}
}

// checks Binacne availability
func (o *BinanceAPIOrchestrator) checkAPIStatusLoop(ctx context.Context, cooldown int) {
	ticker := time.NewTicker(time.Duration(cooldown) * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			response, err := o.binanceAPI.Get(ctx, "/v3/ping", endpoints["/v3/ping"], nil) // get Binance accessibility
			if err != nil {
				o.checkBinanceResponse(nil)
			} else {
				o.checkBinanceResponse(response)
			}
			if o.isBinanceOnline {
				return
			}
		}
	}
}

func (o *BinanceAPIOrchestrator) updateTickerCurrentPrice(ctx context.Context, cooldown int) {
	ticker := time.NewTicker(time.Duration(cooldown) * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			fmt.Println("updateTickerCurrentPrice running...")
			response, err := o.binanceAPI.Get(ctx, "/v3/ticker/price", endpoints["/v3/ticker/price"], nil) // get ticker price
			if err != nil {
				o.checkBinanceResponse(nil)
				continue
			}
			o.checkBinanceResponse(response)

			currencies, err := GetNeededFieldsFromEndpoint(ctx, "/v3/ticker/price")
			if err != nil {
				appmetrics.ApplicationCriticalErrorsLogging("in updateTickerCurrentPrice - GetNeededFieldsFromEndpoint returned Error", err)
			}
			dataForGraph := extractDataFromTickerCurrentPrice(response, currencies)
			triggeredFormulas := o.DependencyGraph.UpdateVariablesTopologicalKahn(dataForGraph)
			if len(triggeredFormulas) > 0 {
				result, variableValues := o.DependencyGraph.GetFormulasVariables(triggeredFormulas)
				AddTriggerHistory(ctx, result, variableValues)
				repositories.SendPushNotifications(triggeredFormulas, "TEST MESSAGE IN ORCHESTRATOR")
			}
		}
	}
}

func (o *BinanceAPIOrchestrator) updatePriceChange24h(ctx context.Context, cooldown int) {
	ticker := time.NewTicker(time.Duration(cooldown) * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			fmt.Println("updatePriceChange24h running...")
			response, err := o.binanceAPI.Get(ctx, "/v3/ticker/24hr", endpoints["/v3/ticker/24hr"], nil) // get price change
			if err != nil {
				o.checkBinanceResponse(nil)
				continue
			}
			o.checkBinanceResponse(response)

			fields, err := GetNeededFieldsFromEndpoint(ctx, "/v3/ticker/24hr")
			if err != nil {
				appmetrics.ApplicationCriticalErrorsLogging("in updatePriceChange24h - GetNeededFieldsFromEndpoint returned Error", err)
			}
			dataForGraph := extractDataFromPriceChange24h(response, fields)

			triggeredFormulas := o.DependencyGraph.UpdateVariablesTopologicalKahn(dataForGraph)
			if len(triggeredFormulas) > 0 {
				result, variableValues := o.DependencyGraph.GetFormulasVariables(triggeredFormulas)
				AddTriggerHistory(ctx, result, variableValues)
				repositories.SendPushNotifications(triggeredFormulas, "TEST MESSAGE IN ORCHESTRATOR")
			}
		}
	}
}

// проверяет ответы от Binance, попутно вявляя ошибки с его стороны
func (o *BinanceAPIOrchestrator) checkBinanceResponse(response interface{}) {
	if response == nil { // Binance DOWN
		if o.isBinanceOnline {
			appmetrics.AvailabilityMetricEvent(2, 0, "Binance DOWN")
			o.isBinanceOnline = false
		}

		o.mu.Lock()
		defer o.mu.Unlock()

		// отмена всех задач кроме пинга
		for api, cancelFunc := range o.tasks {
			if api != "/v3/ping" {
				cancelFunc()
				delete(o.tasks, api)
				delete(o.taskCooldowns, api)
			}
		}

		// запуск пинга если его нет(не должно быть)
		if _, exists := o.tasks["/v3/ping"]; !exists {
			o.launchAPI(context.Background(), "/v3/ping", 60)
		}
	} else if !o.isBinanceOnline { // Binance UP
		appmetrics.AvailabilityMetricEvent(2, 1, "Binance UP")
		o.isBinanceOnline = true

		o.mu.Lock()
		defer o.mu.Unlock()

		// отмена всех задач
		for _, cancelFunc := range o.tasks {
			cancelFunc()
		}
		o.tasks = make(map[string]context.CancelFunc)
		o.taskCooldowns = make(map[string]int)

		// запуск нужные задач заново
		o.LaunchNeededAPI(context.Background())
	}
}

func extractDataFromPriceChange24h(rawData []byte, fields map[string][]string) map[string]float64 {
	var dataset []map[string]interface{}
	err := json.Unmarshal(rawData, &dataset)
	if err != nil {
		appmetrics.ApplicationCriticalErrorsLogging("ERROR IN extractDataFromPriceChange24h", err)
	}

	datasetDict := make(map[string]map[string]interface{})
	for _, data := range dataset {
		if symbol, ok := data["symbol"].(string); ok {
			datasetDict[symbol] = data
		}
	}

	result := make(map[string]float64)
	for symbol, fieldList := range fields {
		if data, exists := datasetDict[symbol]; exists {
			for _, field := range fieldList {
				if val, ok := data[field]; ok {
					switch v := val.(type) {
					case float64:
						result[symbol+"_"+field] = v
					case float32:
						result[symbol+"_"+field] = float64(v)
					case int:
						result[symbol+"_"+field] = float64(v)
					case int64:
						result[symbol+"_"+field] = float64(v)
					case string:
						if parsedVal, err := strconv.ParseFloat(v, 64); err == nil {
							result[symbol+"_"+field] = parsedVal
						}
					}
				}
			}
		}
	}
	return result
}

func extractDataFromTickerCurrentPrice(rawData []byte, currencies map[string][]string) map[string]float64 {
	var dataset []map[string]interface{}
	err := json.Unmarshal(rawData, &dataset)
	if err != nil {
		appmetrics.ApplicationCriticalErrorsLogging("ERROR IN extractDataFromPriceChange24h", err)
	}
	datasetDict := make(map[string]map[string]interface{})
	for _, data := range dataset {
		if symbol, ok := data["symbol"].(string); ok {
			datasetDict[symbol] = data
		}
	}

	result := make(map[string]float64)
	for symbol := range currencies {
		if data, exists := datasetDict[symbol]; exists {
			if priceVal, ok := data["price"]; ok {
				switch v := priceVal.(type) {
				case float64:
					result[symbol+"_price"] = v
				case float32:
					result[symbol+"_price"] = float64(v)
				case int:
					result[symbol+"_price"] = float64(v)
				case int64:
					result[symbol+"_price"] = float64(v)
				case string:
					if parsedVal, err := strconv.ParseFloat(v, 64); err == nil {
						result[symbol+"_price"] = parsedVal
					}
				}
			}
		}
	}
	return result
}
